### ISSUE TITLE
Flickering LEDs while DIM_UP and DIM_DOWN functions is called

### ISSUE THREAD

**Issue by MagnusPer** [Adminius/DimmerControl#5]
When using the One_button_dimmer example the LEDs are flickering while the DIM_UP or DIM_DOWN function is called.  Trying to understand how to change behavior to get a smoother dimmer without flickering.

I suppose the _updatedInterval in the DimmerControl.cpp handle this?
`_updateInterval = (byte)((_valueMax - _valueMin)/ 10) + 1;`

My configuration today is,  valueMax = 100 and valueMin = 0. Meaning (100-0)/10+1 = 9.
But what does that really mean in terms of update intervall frequency?

**Comment by Adminius** [comment]
*_updateInterval* handles **updateAvailable()** only.
and **updateAvailable()** is only for notification purpose (e.g. Serial.print).
It can't be a reason for your problem.
I've tested WS2812B with this lib. all works perfect from about 10% to 100%. but 0% - 9% WS2812B flickers. Then I've tried to test WS2812B without this lib and they flickering also (i've tried to blink between 7% and 8%).
Flickring is also depends on PWM frequency, dimming curve (your look up array) and hardware (e.g. MOSFET+LED combination).

Pls show your sketch, so I can check it.

**Comment by MagnusPer** [comment]
Thanks for helping out.
The flickering is during the complete dim sequence and not specific in the lower parts. One idea could be that I only have 100 values instead of 255, that could cause delays?

It does not flickering for the functions DIM_SOFTON or DIM_SOFTOFF 

I using a NodeMCU standard PWM frequency of 1kHz, have not tested to increased this yet.

[Homey_LedDimmer_Code.zip](https://github.com/Adminius/DimmerControl/files/2626621/Homey_LedDimmer_Code.zip)

[LED_Dimmer_Sketch.zip](https://github.com/Adminius/DimmerControl/files/2626623/LED_Dimmer_Sketch.zip)

**Comment by Adminius** [comment]
i've not tested yet, but I see the first problem: delay(500);
led.task(); should be called as often as possible, without any delays.
in you your case, you have some times 1 or 2 500ms delays -> up to 1 second long LED will not be dimmed.
Try to put led.task() into yield() function (not tested yet)

**Comment by MagnusPer** [comment]
The delays you see is only executed during startup/Power on the first ~10sec till the connection to Homey is establshed. This in order sync the OFF state. Secondly when the LEDs is triggered ON from the Homey. I do not have any flickering at all when controlling the dim levels remotely, only when using the push button

Have now read about yeild(), should I write `yield(led.task());` ?

**Comment by Adminius** [comment]
This way:
```
void yield(void) {
  led.task();
}
```
you have delays in the loop-function:
```
if(SendOnOff == false or SendDim == false){
...
   delay(500);
}
```
and
```
if (SendDimInLoop){   
      delay(500);
...
}
```

**Comment by MagnusPer** [comment]
Thanks!
I will do tests to remove the delays and include the yield function, I will feedback you...

**Comment by MagnusPer** [comment]
Feedback about my tests to lower the LED flicker.
I did the following tests:
1. Removed delays in loop() -  No success, same flickering!
2. Used the yeild() with led.task() within - No success, same flickering!
4. Increased the PWM frequency to 2kHz - No success, same flickering!
5. Decresed the `led.setDurationRelative(4000);` to 2000 - As smother dim cure starting to be visuable!
3. Incresed the `led.setMaxValue(100);` to 255 and also 1023 (standard esp8266) - Now the flickering was much better and only visuable in the lower dim region (0-10%)

**Conslusion** 
The flickering was mainly because of to the low amount of steps (100) used. My setup will be the following  
`led.setMaxValue(1023);`
`led.setDurationRelative(4000);`
